pruner/pruner.go

```diff
@@ -0,0 +1,339 @@
+// automatically triggered during standard operations.
+//
+// Currently an AddPattern, DelPattern, or MatchesForJSONEvent can
+// trigger a rebuild.  When a rebuild is triggered, it's executed
+// synchronous, the the Add/Del method doesn't return until the
+// rebuild is complete.
+type rebuildTrigger interface {
+	// Rebuild should return true to trigger a rebuild.
+	//
+	// This method is called by AddPattern and DelPattern.  added
+	// is true when called by AddPattern; false otherwise. These
+	// methods do not return until the rebuild is complete, so
+	// beware.
+	Rebuild(added bool, s *Stats) bool
+}
+
+// NewMatcher does what you'd expect.
+//
+// The State defaults to MemState.
+func NewMatcher(s State) *Matcher {
+	if s == nil {
+		s = NewMemState()
+	}
+	return &Matcher{
+		Matcher:        quamina.NewMatcher(),
+		live:           s,
+		rebuildTrigger: defaultRebuildTrigger,
+	}
+}
+
+// maybeRebuild calls rebuildTrigger (if not) and then calls rebuild()
+// if that trigger said to do that.  If rebuildTrigger is nil, no
+// rebuild is executed.
+//
+// This method assumes the caller has a write lock.
+func (m *Matcher) maybeRebuild(added bool) error {
+	if m.rebuildTrigger == nil {
+		return nil
+	}
+	if m.rebuildTrigger.Rebuild(added, &m.stats) {
+		return m.rebuild(added)
+	}
+
+	return nil
+}
+
+// AddPattern calls the underlying AddPattern method and then maybe
+// rebuilds the index (if the AddPattern succeeded).
+func (m *Matcher) AddPattern(x quamina.X, pat string) error {
+	var err error
+
+	// Do we m.live.Add first, do we m.Matcher.AddPattern first?
+	if err = m.Matcher.AddPattern(x, pat); err == nil {
+		m.lock.Lock()
+		m.stats.Added++
+		m.stats.Live++
+		m.maybeRebuild(true)
+		m.lock.Unlock()
+		err = m.live.Add(x, pat)
+		// ToDo: Contemplate what do to about an error here
+		// (or if we got an error from AddPattern after we did
+		// live.Add.
+	}
+
+	return err
+}
+
+// NewFJ just calls quamina.FJ.
+func NewFJ(m *Matcher) quamina.Flattener {
+	return quamina.NewFJ(m.Matcher)
+}
+
+// NewFJ calls quamina.NewFJ with this Matcher's core quamina.Matcher
+func (m *Matcher) NewFJ() quamina.Flattener {
+	return quamina.NewFJ(m.Matcher)
+}
+
+// MatchesForJSONEvent calls the underlying MatchesForJSONEvent method
+// and then maybe rebuilds the index.
+func (m *Matcher) MatchesForJSONEvent(event []byte) ([]quamina.X, error) {
+	fs, err := m.NewFJ().Flatten(event)
+	if err != nil {
+		return nil, err
+	}
+	return m.MatchesForFields(fs)
+}
+
+func (m *Matcher) MatchesForFields(fields []quamina.Field) ([]quamina.X, error) {
+
+	xs := m.Matcher.MatchesForFields(fields)
+
+	// Remove any X that isn't in the live set.
+
+	acc := make([]quamina.X, 0, len(xs))
+
+	// We're updating stats.Filtered, so we need a write lock.  If
+	// we forget about stats.Filtered, we can live with only a
+	// read lock here.
+	var emitted, filtered int64
+	for _, x := range xs {
+		have, err := m.live.Get(x)
```

Comment:
could the variable names be a bit more descriptive?  I think "have" means something like "filterThisMatch" or "patternIsDeleted"

Reply:
Ah, I missed this comment, too. See above re [short variable name style](https://github.com/golang/go/wiki/CodeReviewComments#variable-names). But still yeah, I can do that.